File: solutions/fsharp/sieve/Sieve.fs
module Sieve

let primes limit =
    let rec sieve numbers primes =
        match numbers with
        | [] -> List.rev primes
        | prime :: rest ->
            let multiples = [ prime * prime .. 2 * prime .. limit ]
            sieve (rest |> List.except multiples) (prime :: primes)

    match limit < 2 with
    | true -> []
    | _ -> sieve [ 3..2..limit ] [ 2 ]

File: solutions/elixir/sieve/lib/sieve.ex
defmodule Sieve do
  @doc """
  Generates a list of primes up to a given limit.
  """
  @spec primes_to(non_neg_integer) :: [non_neg_integer]
  def primes_to(limit) when limit < 2, do: []

  def primes_to(limit) do
    Enum.to_list(2..limit)
    |> find_primes(limit, [])
    |> Enum.reverse()
  end

  @spec find_primes([integer], integer, [integer]) :: [integer]
  defp find_primes([], _, primes), do: primes

  defp find_primes([prime | rest], limit, primes) do
    (rest -- Enum.to_list((prime * prime)..limit//prime))
    |> find_primes(limit, [prime | primes])
  end
end
